File: ios/Classes/observers/ChimeDataMessageObserver.swift
//
//  ChimeDataMessageObserver.swift
//  eggnstone_amazon_chime
//
//  Created by 長谷川 樹 on 2022/07/25.
//

import Foundation
import AmazonChimeSDK
import Flutter

public class ChimeDataMessageObserver : DataMessageObserver {
    let _eventSink: FlutterEventSink
    
    init(eventSink: @escaping FlutterEventSink) {
        self._eventSink = eventSink
    }
    
    public func dataMessageDidReceived(dataMessage: DataMessage) {
        let arguments = """
            {
            "Data": "\(dataMessage.data)",
            "SenderAttendeeId": "\(dataMessage.senderAttendeeId)",
            "SenderExternalUserId": "\(dataMessage.senderExternalUserId)",
            "Text": \(dataMessage.text()!),
            "Throttled": "\(dataMessage.throttled)",
            "TimestampMs": "\(dataMessage.timestampMs)",
            "Topic": "\(dataMessage.topic)"
            }
            """

        let json = """
            {
            "Name": "OnDataMessageDidReceived",
            "Arguments": \(arguments)
            }
            """

        _eventSink(json)
    }
}

File: android/src/main/kotlin/dev/eggnstone/chime/observers/ChimeDataMessageObserver.kt
package dev.eggnstone.chime.observers
import com.amazonaws.services.chime.sdk.meetings.realtime.datamessage.DataMessage
import com.amazonaws.services.chime.sdk.meetings.realtime.datamessage.DataMessageObserver
import io.flutter.plugin.common.EventChannel.EventSink
import org.json.JSONObject

class ChimeDataMessageObserver(private val _eventSink: EventSink) : DataMessageObserver
{
    override fun onDataMessageReceived(dataMessage: DataMessage)
    {
        val jsonObject = JSONObject()
        val eventArguments = JSONObject()
        eventArguments.put("TimestampMs", dataMessage.timestampMs)
        eventArguments.put("Topic", dataMessage.topic)
        eventArguments.put("SenderAttendeeId", dataMessage.senderAttendeeId)
        eventArguments.put("SenderExternalUserId", dataMessage.senderExternalUserId)
        eventArguments.put("Throttled", dataMessage.throttled)
        eventArguments.put("Data", dataMessage.data)
        eventArguments.put("Text", dataMessage.text())
        jsonObject.put("Name", "onDataMessageReceived")
        jsonObject.put("Arguments", eventArguments)
        _eventSink.success(jsonObject.toString())
    }
}
